### PR TITLE
Enable launching an app without attaching to a WDA or XCUITest session

### PR DESCRIPTION
**Pull request by jobeylev**
The /wda/apps/launch endpoint registers the launched app with the active session, and uses XCUIApplication to perform the launch. This causes the app to be automatically terminated (by both WDA and XCUITest) when WDA terminates. I added the endpoint wda/launchUnattachedApp to enable launching an app in such a way that it remains unaffected by WDA terminating. This is critical for Testim's use of WDA, which includes frequently re-running WDA while keeping the application under test open.

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
The CLA bot checks the email, which is set for your local GitHub profile. Make sure it is the same, as the email provided to the CLA page

File: PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
+//
```
Comment (mykola-mokhnach) [inline PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h]: Please add the standard facebook copyright header

File: PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+/* Generated by RuntimeBrowser
+ Image: /System/Library/Frameworks/MobileCoreServices.framework/MobileCoreServices
+ */
+
+@interface LSApplicationWorkspace : NSObject {
+  NSXPCConnection * _connection;
+  NSMutableDictionary * _createdInstallProgresses;
+//  LSInstallProgressDelegate * _delegateProxy;
```
Comment (mykola-mokhnach) [inline PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h]: please remove the commented stuff if we don't need it

File: WebDriverAgentLib/Commands/FBCustomCommands.m [hunk 3]
```diff
@@ -55,6 +56,7 @@ + (NSArray *)routes
 #endif
     [[FBRoute POST:@"/wda/pressButton"] respondWithTarget:self action:@selector(handlePressButtonCommand:)],
     [[FBRoute POST:@"/wda/siri/activate"] respondWithTarget:self action:@selector(handleActivateSiri:)],
+    [[FBRoute POST:@"/wda/launchUnattachedApp"].withoutSession respondWithTarget:self action:@selector(handleLaunchUnattachedApp:)],
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBCustomCommands.m]: there is already an endpoint to launch apps. Can we just add another option to it?

File: PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h (hunk 1 above)
Comment (jobeylev) [reply PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h]: I am happy to do so if you prefer, but are you sure about adding a Facebook copyright to code that belongs to Apple?

File: PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h (hunk 2 above)
Comment (jobeylev) [reply PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h]: Done

File: PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h (hunk 1 above)
Comment (mykola-mokhnach) [reply PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h]: then just remove the whole comment and only keep the `generated` note

File: WebDriverAgentLib/Commands/FBCustomCommands.m [hunk 4]
```diff
@@ -227,4 +229,20 @@ + (BOOL)isKeyboardPresent {
...
 
++ (id <FBResponsePayload>)handleLaunchUnattachedApp:(FBRouteRequest *)request
+{
+  NSString *bundle = (NSString *)request.arguments[@"bundleId"];
+  if ([[LSApplicationWorkspace defaultWorkspace] openApplicationWithBundleID:bundle])
+    return FBResponseWithOK();
+  return FBResponseWithError([self appLaunchError]);
+}
+
++ (NSError *)appLaunchError
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBCustomCommands.m]: do we really need a separate method to build an error? Also, WDA has an error builder class

**Comment by mykola-mokhnach** [comment]
It would be good to have an integration test for the newly added functionality

File: WebDriverAgentLib/Commands/FBCustomCommands.m (hunk 3 above)
Comment (imurchie) [reply WebDriverAgentLib/Commands/FBCustomCommands.m]: A duplicate endpoint with `withoutSession` is possible, too, as in [lock](https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Commands/FBCustomCommands.m#L42-L43)

File: WebDriverAgentLib/Commands/FBCustomCommands.m (hunk 4 above)
Comment (KazuCocoa) [reply WebDriverAgentLib/Commands/FBCustomCommands.m]: WDA has an error builder class => `FBErrorBuilder`

**Review by KazuCocoa — COMMENTED** [review]
👍

File: PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h (hunk 1 above)
Comment (jobeylev) [reply PrivateHeaders/MobileCoreServices/LSApplicationWorkspace.h]: Done

File: WebDriverAgentLib/Commands/FBCustomCommands.m (hunk 3 above)
Comment (jobeylev) [reply WebDriverAgentLib/Commands/FBCustomCommands.m]: I think @imurchie makes a good point: this functionality belongs in a `withoutSession` endpoint.      However, don't you think this is not explicit enough? How would a user guess that passing the `session` argument into `wda/apps/launch` will result in the app dying when WDA dies, but that not passing the `session` will result in the app surviving the death of WDA? On the other hand, if we take another argument into `wda/apps/launch`, then won't we have a kind of duplication? In order to launch an unattached app the user will have to both pass `unattached = true` and not pass `session`. I'm happy to do what you think is best.

File: WebDriverAgentLib/Commands/FBCustomCommands.m (hunk 4 above)
Comment (jobeylev) [reply WebDriverAgentLib/Commands/FBCustomCommands.m]: Fixed

**Comment by jobeylev** [comment]
@mykola-mokhnach , apologies but I need a hint about how to create the integration test. Specifically, how should I pass the `wda/launchUnattachedApp` message to the server in the integration test? I cannot find an example of a test that does that. Can I assume the server will be on a given IP? Or is there a better way?

**Comment by mykola-mokhnach** [comment]
@jobeylev Integration tests are located in the separate folder. Here is an example:
https://github.com/appium/WebDriverAgent/blob/0fc5ae36726eae4798ff0543f5edebb9ad91361d/WebDriverAgentTests/IntegrationTests/FBSessionIntegrationTests.m#L37

In your case you could simply call `[[LSApplicationWorkspace defaultWorkspace] openApplicationWithBundleID:bundle]` in you test and then assert the app is running.

File: WebDriverAgentLib/Commands/FBCustomCommands.m (hunk 3 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Commands/FBCustomCommands.m]: I got you point. What if we rename the endpoint to `wda/apps/launchUnattached` to just follow the common approach?

Comment (jobeylev) [reply WebDriverAgentLib/Commands/FBCustomCommands.m]: Done

**Comment by jobeylev** [comment]
@mykola-mokhnach , I added an integration test that verifies the app is launched and is not attached to the session.

File: WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+//  Copyright © 2019 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FBUnattachedAppLauncher : NSObject
+
++ (BOOL)launchAppWithBundleId:(NSString *)bundleId;
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h:23]: Please add a docstring

File: WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
+//
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h]: please fix the header

File: WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.m [hunk 7]
```diff
@@ -0,0 +1,18 @@
+//
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.m]: please fix the header

File: WebDriverAgentTests/IntegrationTests/FBSessionIntegrationTests.m [hunk 8]
```diff
@@ -89,4 +95,12 @@ - (void)testMainAppCanBeRestartedInScopeOfTheCurrentSession
   XCTAssertEqualObjects(testedApp.bundleID, self.session.activeApplication.bundleID);
 }
 
+- (void)testLaunchUnattachedApp
```
Comment (mykola-mokhnach) [inline WebDriverAgentTests/IntegrationTests/FBSessionIntegrationTests.m:98]: 👍

File: WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h (hunk 5 above)
Comment (jobeylev) [reply WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h:23]: Done

File: WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h (hunk 6 above)
Comment (jobeylev) [reply WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.h]: Done

File: WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.m (hunk 7 above)
Comment (jobeylev) [reply WebDriverAgentLib/Utilities/FBUnattachedAppLauncher.m]: Done

**Review by KazuCocoa — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Waiting for CI to be completed

**Comment by mykola-mokhnach** [comment]
> https://travis-ci.org/appium/WebDriverAgent/jobs/556305184#L1237

It looks like the new command requires ifdef for appletv platform

**Comment by jobeylev** [comment]
It seems to cause was I was not aware of the tvOS target, and so didn't add the new class I created to it. Should be fixed now. Let's see what Travis says.

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
CI errored out due to some internal xcode issue, this can be ignored

**Pull request merged**